procrastinate/sql/migrations/02.00.03_01_add_job_priority.sql

```diff
@@ -0,0 +1,124 @@
+ALTER TABLE procrastinate_jobs ADD COLUMN priority integer DEFAULT 0 NOT NULL;
+
+CREATE UNIQUE INDEX idx_procrastinate_jobs_priority_id ON procrastinate_jobs (priority, id) WHERE status = 'todo';
+
+DROP FUNCTION IF EXISTS procrastinate_defer_job(character varying, character varying, text, text, jsonb, timestamp with time zone);
+CREATE OR REPLACE FUNCTION procrastinate_defer_job(
```

Comment:
Do we need the `DROP` if we use `CREATE OR REPLACE` ? It's very possible that we do, there's something in the back of my mind telling me there's a gotcha in the `CREATE OR REPLACE` thing, and I can't remember what it is.

Reply:
I just did it analogous to the existing migrations, but I was also wondering. But it should at least not break something.

Reply:
Ok, re-read the [Postgres doc](https://www.postgresql.org/docs/current/sql-createfunction.html). `CREATE OR REPLACE` won't let you change the number or types of arguments or the return type.

It's also noted that if we drop and recreate a function, we need to recreate the corresponding triggers, I wasn't aware of that so it's a good thing I got to re-read that :) The following functions would be impacted: `procrastinate_notify_queue`, `procrastinate_trigger_status_events_procedure_update`, `procrastinate_trigger_status_events_procedure_insert`, `procrastinate_trigger_scheduled_events_procedure`, `procrastinate_unlink_periodic_defers`. I think we never needed to change those, so never got to trigger a bug with that, but I'll know for the future.

In our case it... doesn't hurt to use CREATE OR REPLACE, even if there shouldn't be a case where it's replaced.

Also, there's the question of backwards compatibility. The [rule we have in place](https://github.com/procrastinate-org/procrastinate/blob/main/CONTRIBUTING.md#backward-compatibility) says you should:
- Not delete the old function (so remove the DROP statement)
- Keep a copy of both the old and new functions in schema.sql (the old ones are at the bottom)
- If you want, prepare a future migration in `procrastinate/sql/future_migrations/03.00.00_01_drop_old_functions_before_job_priority.sql` that contains the DROP statements. It will be integrated next time we do a major version (if we ever do one).

Reply:
I think the backwards compat element is the only thing left before we can merge ? I'll re-read and double check.

Reply:
Thanks for the reminder. I'll look at the backward-compatible migrations later (or maybe tomorrow). I'm not exactly sure about it, but I will figure it out.

Reply:
Honnestly, I think just not deleting the old functions and keeping both old and new functions in schema.sql should be enough for the functions that change signature. 

For the functions that don't change signature, you probably don't have anything to do.


Keeping in mind that the new priority column has a default and if every task has priority zero, then it works the same as before, I don't foresee specific issues.

Reply:
So removing this line:
https://github.com/procrastinate-org/procrastinate/pull/1070/files#diff-78aec0cd2e7f3a027ae5711665e34f4068d8bea1ecfaeca5cd35afca2c26ba76R3

And duplicating the `procrastinate_defer_job` function in schema.sql so that both the old function without the pirority param and the new function with the priority param exist (put the legacy function at the bottom)

Reply:
Ah, now I get it. It works because PostgreSQL supports function overloading. I hope it's ok now.

By the way ... shouldn't the old `procrastinate_finish_job` function shouldn't be removed by now? Above the old function, it says "to be removed after 2.0.0," and above the new "it is to be removed after 1.0.0 is released". Is it ok if I clean the comments up a bit (maybe just add at the "Old versions of functions" title that those will be removed with a version 3 release)?

Reply:
I'll need to double check, this is probably just the comment that's outdated. I'll check outside of the PR.